python-avd/pyavd/_cv/workflows/verify_inputs.py

```diff
@@ -0,0 +1,203 @@
+        devices=devices
+    )
+
+    if duplicated_serial_number or duplicated_system_mac_address_unset_serial_number or duplicated_system_mac_address_set_serial_number:
+        duplicated_devices_handler(
+            duplicated_serial_number=duplicated_serial_number,
+            duplicated_system_mac_address_unset_serial_number=duplicated_system_mac_address_unset_serial_number,
+            duplicated_system_mac_address_set_serial_number=duplicated_system_mac_address_set_serial_number,
+            strict_system_mac_address=strict_system_mac_address,
+            warnings=warnings,
+        )
+
+
+def identify_duplicated_devices(
+    *, devices: list[CVDevice]
+) -> tuple[list[dict[str, str | list[CVDevice]]], list[dict[str, str | list[CVDevice]]], list[dict[str, str | list[CVDevice]]]]:
+    """
+    Process list of CVDevice instances to identify those with overlapping serial_number or system_mac_address.
+
+    Return tuple containing:
+      - List of dictionaries with information about CVDevices with overlapping serial_number.
+      - List of dictionaries with information about CVDevices with overlapping system_mac_address and unset serial_number.
+      - List of dictionaries with information about CVDevices with overlapping system_mac_address and set serial_number.
+    """
+    # List holding CVDevices with duplicated serial_number
+    duplicated_serial_number: list[dict[str, str | list[CVDevice]]] = []
+    # List holding CVDevices with duplicated system_mac_address and unset
+    duplicated_system_mac_address_unset_serial_number: list[dict[str, str | list[CVDevice]]] = []
+    # List holding CVDevices with duplicated system_mac_address
+    duplicated_system_mac_address_set_serial_number: list[dict[str, str | list[CVDevice]]] = []
+
+    # Deduplicate CVDevice objects as original `devices` list may contain duplicated items
+    unique_devices = list({id(device): device for device in devices}.values())
+
+    # Group devices based on <CVDevice>.serial_number as long as it's not None
+    devices_grouped_by_serial_number = groupby_obj(
+        list_of_objects=[device for device in unique_devices if device.serial_number is not None], attr="serial_number"
+    )
+
+    # Group devices based on <CVDevice>.system_mac_address as long as it's not None
+    devices_grouped_by_system_mac_address = groupby_obj(
+        list_of_objects=[device for device in unique_devices if device.system_mac_address is not None], attr="system_mac_address"
+    )
+
+    # Populate list of CVDevice with duplicated serial_number values
+    for current_serial_number, device_iterator_object in devices_grouped_by_serial_number:
+        if len(devices_with_current_serial_number := list(device_iterator_object)) > 1:
+            duplicated_serial_number.append(
+                {"duplicated_serial_number": current_serial_number, "devices_with_duplicated_serial_number": devices_with_current_serial_number}
+            )
```

Comment:
I think you can build this with.
In anycase you don't need the keys "duplicated_serial_number" / "devices_with_duplcyaed_serial_number" as we know what they are in the code so a simple dict being
`duplicated_sn = {sn: devices}` is enough rather than holding maybe 100 times the strings.

```suggestion
    from itertools import filterfalse
    # we can replace lambda with a helper function that takes a tuple (string, list) and returns true or False)
    duplicated_sn = {sn: devices for sn, devices in filterfalse(lambda group: len(group[1]==1, devices_grouped_by_serial_number)}
```

Reply:
hmm scratch the suggestion we return iterators inside the iterator in groupby so not sure `devices` will hold anything in the dict need to test

Reply:
Removed those extra "informative" keys and converted to dicts